### PR TITLE
Fix: restore make_axes to accept a tuple of axes

### PR DESCRIPTION
**Pull request by PiotrStrzelczykX**
## PR Summary

Allow to pass a tuple of axes as "ax" parameter of make_axes function.
It restores functionality available in this function, before memory leak fix
(in PR: https://github.com/matplotlib/matplotlib/pull/22089 ).

## PR Checklist

**Tests and Styling**
- [N/A] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [N/A] New plotting related features are documented with examples.

**Release Notes**
- [N/A] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [N/A] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [N/A] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

**Comment by PiotrStrzelczykX** [comment]
Hello @story645, I added link to PR which changed this behavior.
I didn't prepare issue, because fix is so obvious.
I've spotted a issue after upgrading matplotlib:
```...
AttributeError: 'tuple' object has no attribute 'get_figure' 
Error_place:
  File 'C:\\Python310\\lib\\site-packages\\matplotlib\\colorbar.py':1446 in make_axes
  fig = parents[0].get_figure()
```
 and prepared this fix.

**Comment by jklymak** [comment]
Please add a test and maybe also include your test code in the issue.  If we broke something once it is because there is no test.

**Comment by jklymak** [comment]
Actually I guess we should discuss if we want to accept a tuple.  That may have worked before, but it wasn't documented, or particularly intended.

**Comment by ksunden** [comment]
General typing advice is to be open about what you accept and strict about what you output.

`parents` is used to a) get `parents[0]` and b) loop over (twice)

Perhaps we should be testing/documenting as `collections.abc.Sequence` instead (which would get both list and tuple, as well as anything else which quacks like we use it)?

**Comment by jklymak** [comment]
Sure that sounds reasonable.  However, being permissive is tricky practically when you are trying to write type checks etc, and for testing, since we can't check everything that quacks to make sure we don't break it.

**Comment by tacaswell** [comment]
I suspect the right check here is to see if it is an iterable and special case if it is an array.  Maybe right path is something like

```
if is_iterable(parents):
    if isinstance(parents, np.array):
         list(parents.flat)
    else:
        parents = list(parents)
else:
    parents = [parents]
```

thus we come out of this block it is definitly a list, we special case the np.array case (which might break with some future numpy-clone), and we let Python do the coerces to list for us.

**Comment by PiotrStrzelczykX** [comment]
@tacaswell checking `is_iterable(parents)` is not  firm, as there is also `parents[0]` access, checking `is_instance(parents, collections.abc.Sequence)` (suggested by @ksunden) would be better.

**Comment by PiotrStrzelczykX** [comment]
> Please add a test and maybe also include your test code in the issue. If we broke something once it is because there is no test.

@jklymak test added.

I not insist on this change -- I thought that it may be useful for many users so I push it, but if you think it may be incompatible just discard it.

File: lib/matplotlib/colorbar.py
Comment (rcomer) [inline lib/matplotlib/colorbar.py:1407]:
```suggestion
    parents : `~.axes.Axes` or sequence or `numpy.ndarray` of `~.axes.Axes`
```
?

**Comment by rcomer** [comment]
I think this description in `Figure.colorbar` should also be updated for consistecy
https://github.com/matplotlib/matplotlib/blob/d09526fa50610bf2b3fa5b7077e11679a04b6d00/lib/matplotlib/figure.py#L1197-L1200

**Comment by rcomer** [comment]
@tacaswell’s suggestion of converting any iterable to a list would allow the user to pass `dict.values()`, where `dict` comes from `subplot_mosaic`.  That seems worth having to me.

File: lib/matplotlib/colorbar.py
Comment (PiotrStrzelczykX) [reply lib/matplotlib/colorbar.py]: Thanks. Fixed. (Also in figure.py.)

**Review by rcomer — COMMENTED** [review]
@PiotrStrzelczykX thanks for the update.  The docstrings all look consistent with the change now.

I still think that if we support all sequences, it is worth expanding to support all iterables:

- as @tacaswell notes, containing questions about types in that initial loop and always coming out of that loop with a list is more robust.
- `dict.values()` seems like a likely thing to want to pass (and is not a sequence).
- I suspect that a lot of users would not know off hand the difference between a sequence and an iterable (I always have to look it up!) so it’s just easier from the user point of view if the more general object is supported.

Having said all that, I am just a contributor and have no authority here, so we need to wait for a decision from the core developers.

**Comment by jklymak** [comment]
I think we should actually figure out a consistent way to deal with itterables and perhaps factor it into `_api` so that we do it consistently in different places.  This came up recently in another PR as well (sorry I forget which one)

I wonder if co-ercing to numpy array first is a good consistent way to do this?  Then we are just relying on their broadcasting rules, which is easy to explain.  eg `parents = list(np.asarray(parents).flat)`.  Or we could even just leave it as an array?

**Comment by efiring** [comment]
`list()` does the right thing for dictionaries, while `np.asarray` does not:
```
In [3]: d = dict(a=1, b=2)

In [4]: import numpy as np

In [5]: np.asarray(d.values())
Out[5]: array(dict_values([1, 2]), dtype=object)

In [6]: list(d.values())
Out[6]: [1, 2]
```

**Comment by ksunden** [comment]
> @tacaswell checking `is_iterable(parents)` is not firm, as there is also `parents[0]` access, checking `is_instance(parents, collections.abc.Sequence)` (suggested by @ksunden) would be better.

The code provided by @tacaswell enforces that it is a `list` (which can be indexed) when `parents[0]` is called, so it is perfectly fine for any (finite) iterable. So that code is in fact more general, and will accept more inputs as valid.

**Comment by efiring** [comment]
@PiotrStrzelczykX, thank you for this PR.  In fact, I tripped over this (not accepting a tuple argument) myself a week or two ago.  We will be happy to accept the PR *either* as-is, *or* after one more change, if you choose to make it: specifically, the suggestion by @tacaswell, and endorsed by @rcomer and myself, to broaden the supported types to include iterables, like `dict.values()`.  Fleshed out, it becomes
```python
if np.iterable(parents):
    if isinstance(parents, np.ndarray):
         parents = list(parents.flat)
    else:
        parents = list(parents)
else:
    parents = [parents]
```
If you do choose to make this change, then where you have "sequence" in the docstring it would become the more general "iterable".

It is possible that some time later, after merging the PR, we might factor this argument-handling functionality out for use in other places; but we don't want to stall the PR while hashing out larger questions of API design and factoring.

**Comment by efiring** [comment]
Also, if you choose to make the change, you could add a line to the test:
```python
fig.colorbar(im, ax=dict(a=ax[0], b=ax[1]).values())
```

**Comment by rcomer** [comment]
I just noticed that the constrained layout guide states
> If you specify a list of axes (or other iterable container) to the ax argument of colorbar, constrained_layout will take space from the specified axes.

https://matplotlib.org/stable/tutorials/intermediate/constrainedlayout_guide.html#colorbars

@PiotrStrzelczykX could you let us know if you intend to take this any further?  As @efiring said, we are happy to merge as is.  However, it would be better not to merge yet if you are planning to push further changes.

**Review by tacaswell — APPROVED** [review]

**Comment by tacaswell** [comment]
I'm going to merge this on my approval and the comments from @efiring and @rcomer  and then open up a follow on PR with the additional changes.

**Comment by tacaswell** [comment]
Thank you for your work on this @PiotrStrzelczykX and congratulations on your first merged Matplotlib PR 🎉  I hope we hear from you again!

**Pull request merged**
